docs/irma-protocol.md

```diff
@@ -0,0 +1,376 @@
+      },
+      "a_disclosed": {
+        "1": "AwAKhwAaAAXZZxdMn4TvQ6F/mVxWb6a7",
+        "2": "5MrO6tjC5Q=="
+      }
+    }
+  ],
+  "indices": [
+    [
+      {
+        "cred": 0,
+        "attr": 2
+      }
+    ]
+  ]
+}
+```
+
+The `proofs` array contains a list of disclosure proofs, i.e. [`gabi.ProofD`](https://pkg.go.dev/github.com/privacybydesign/gabi#ProofD) instances, for each credential out of which attributes are being disclosed (present in the `a_disclosed` map, see below). The remainder of the fields within the `gabi.ProofD` are [zero-knowledge proofs](zkp.md) of the hidden attributes (`a_responses`) and of the issuer signature over the attributes of this credential.
+
+The `indices` double array informs the IRMA server where it can find the attributes that it requested in the session request within the `proofs` array, schematically as follows: if `indices[i][j]` contains `{"cred": k, "attr": l}`, then the attribute requested in the [inner conjunction](session-requests.md#multiple-credential-types-within-inner-conjunctions) `j` in the outer conjunction `i` can be found in `proofs[k].a_disclosed[l]`.
+
+#### The attributes in `a_disclosed`
+
+The IRMA protocol uses Base64 to encode large integers, as can be seen in the strings in the message above. The disclosed attributes additionally contain an extra bit in the least significant position encoding their presence, which is used to distinguish [absent attributes](session-requests.md#issuance-requests) from attributes containing the empty string as values. Thus, a (non-empty) string found within `a_disclosed` containing a disclosed attribute can be converted to its string value as follows:
+
+* Base64-decode the string,
+* Bit-shift the resulting bytes one position to the right,
+* Parse the resulting bytes as a UTF-8 string.
+
+For example, taking attribute `2` from `a_disclosed` in the message above, the following [Go snippet](https://play.golang.org/p/GTWeM_uHr98) outputs `regular`.
+
+```go
+bts, _ := base64.StdEncoding.DecodeString("5MrO6tjC5Q==")
+i := new(big.Int).SetBytes(bts)
+i.Rsh(i, 1)
+fmt.Println(string(i.Bytes()))
+```
+
+Note that attribute `1` is the [metadata attribute](overview.md#the-metadata-attribute), containing among others the credential type and the expiry date of the credential in a custom encoding. This attribute is always disclosed. The above snippet will not output anything sensible for metadata attributes, but instead the [`irma` command line tool](irma-cli.md) can be used as follows.
+
+```text
+$ irma meta "AwAKhwAaAAXZZxdMn4TvQ6F/mVxWb6a7"
+Identifier      : pbdf.pbdf.irmatube
+Signed          : 2021-08-26 02:00:00 +0200 CEST
+Expires         : 2022-02-24 01:00:00 +0100 CET
+IsValid         : true
+Version         : 3
+KeyCounter      : 5
+KeyExpires      : 2021-09-23 11:43:09 +0200 CEST
+KeyModulusBitlen: 2048
+```
+
+### ABS: `POST /irma/session/{clientToken}/proofs`
```

Comment:
I have expanded the abbreviation.

> It also looks a bit weird that we have two sections now for the proofs endpoint.

I kind of agree. I did it like this anyway to get the endpoints to show up in the table of contents shown on the right hand side of the page (see screenshot). I like having the endpoints show up there, and the other pages documenting HTTP endpoints have that too.

<img width="238" alt="Screenshot 2021-11-03 at 08 31 05" src="https://user-images.githubusercontent.com/979897/140028868-90784895-4b0d-487e-82f0-c574dcd6ee29.png">

It is unfortunately not an option to use an extra level in the sections to solve this (i.e. #### instead of ###), because deeper levels don't show up in the TOC there. So besides removing all endpoints from the section headers/TOC, I don't see a nice way to solve this.

Reply:
I see, I had this problem before too. A bit unfortunate but it's as good as it gets then now.